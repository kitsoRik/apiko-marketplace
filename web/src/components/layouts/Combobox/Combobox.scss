.combobox {
	position: relative;

	&-options {
		width: 100%;
		position: absolute;
		top: 100%;

		z-index: 2500;
	}

	&::before, &::after {
		content: " ";
		position: absolute;
		top: calc(50% - 7px);
		right: calc(7px + 6px);
		height: 15px;
		width: 2px;

		border-radius: 122px;

		transform: rotate(45deg);

		background-color: black;
		pointer-events: none;
	}

	&::after {
		right: calc(7px + 1px + 15px);
		transform: rotate(-45deg);
	}
}
