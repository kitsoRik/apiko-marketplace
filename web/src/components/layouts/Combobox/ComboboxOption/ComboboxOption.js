import React from "react";

import "./ComboboxOption.scss";

const ComboboxOption = ({ value, icon, children, ...props }) => {
	return (
		<div className="combobox-option" {...props}>
			<div className="combobox-option-icon">{icon}</div>
			<p className="combobox-option-text">{children}</p>
		</div>
	);
};

export default ComboboxOption;
