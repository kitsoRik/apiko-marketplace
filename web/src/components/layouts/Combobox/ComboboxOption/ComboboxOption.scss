.combobox-option {
	padding: 8px;
	width: 100%;

	background-color: #f9fafb;
	color: #505050;

	border: 1px solid #dedee0;

	display: grid;
	grid-template-columns: 18px auto 15px;
	grid-column-gap: 10px;

	appearance: none;
	outline: none;

	border-radius: 5px;
	box-sizing: border-box;

	cursor: pointer;

	user-select: none;

	&:hover {
		background-color: #e8e9ea;
	}

	&-icon {
		height: 18px;
		width: 18px;
	}

	&-text {
		white-space: nowrap;
		overflow: hidden;
		text-overflow: ellipsis;
	}
}
