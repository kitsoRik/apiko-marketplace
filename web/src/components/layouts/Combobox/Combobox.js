import React, { useState } from 'react';

import "./Combobox.scss";
import ComboboxOption from './ComboboxOption/ComboboxOption';

const Combobox = ({ value, children = [], onChange, ...props }) => {

    const [dropable, setDropable] = useState(false);


    const onClick = (e) => {
        setDropable(!dropable);
    }

    const onClickOption = (value) => {
        if (onChange)
            onChange(value);
        setDropable(false);
    }

    let optionsElements = [];

    children.forEach(c => {
        if (c.type === ComboboxOption) optionsElements.push(c);
    });

    let select = null;

    optionsElements = React.Children.map(optionsElements, c => {
        if (c.props.value === value) {
            select = React.cloneElement(c, { ...c.props, onClick: onClick });
        }
        return React.cloneElement(c, { ...c.props, onClick: () => onClickOption(c.props.value) })
    });


    return (
        <div tabIndex={0} className="combobox" onBlur={() => setDropable(false)} {...props}>
            {select}
            {dropable && <div className="combobox-options">
                {
                    optionsElements
                }
            </div>
            }
        </div>
    );
}

export default Combobox;