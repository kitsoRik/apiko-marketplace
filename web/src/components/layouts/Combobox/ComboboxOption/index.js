import ComboboxOption from "./ComboboxOption";

export default ComboboxOption;
