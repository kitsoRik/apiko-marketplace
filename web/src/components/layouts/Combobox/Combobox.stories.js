import React from 'react';
import Combobox from './Combobox';
import ComboboxOption from './ComboboxOption/ComboboxOption';

export const medium = () => 
<Combobox type="medium">
    <ComboboxOption value="1">1</ComboboxOption>
    <ComboboxOption value="2">2</ComboboxOption>
</Combobox>

export default { title: "Combobox" }