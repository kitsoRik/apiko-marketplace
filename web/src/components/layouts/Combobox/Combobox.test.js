import React from "react";
import { create } from "react-test-renderer";
import Combobox from "./Combobox";
import ComboboxOption from "./ComboboxOption/ComboboxOption";

describe("Combobox", () => {
	it("should render without options", () => {
		const component = create(<Combobox></Combobox>);
		const json = component.toJSON();

		expect(json).toMatchSnapshot();
	});

	it("should render with options", () => {
		const component = create(
			<Combobox value="1">
				<ComboboxOption value="1">Hello</ComboboxOption>
				<ComboboxOption value="2">World</ComboboxOption>
			</Combobox>
		);
		const json = component.toJSON();

		expect(json).toMatchSnapshot();
	});
});
