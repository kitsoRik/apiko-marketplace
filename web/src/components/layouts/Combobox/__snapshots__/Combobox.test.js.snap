// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox should render with options 1`] = `
<div
  className="combobox "
  onBlur={[Function]}
  tabIndex={0}
>
  <div
    className="combobox-option"
    onClick={[Function]}
  >
    <div
      className="combobox-option-icon"
    />
    <p
      className="combobox-option-text"
    >
      Hello
    </p>
  </div>
</div>
`;

exports[`Combobox should render without options 1`] = `
<div
  className="combobox "
  onBlur={[Function]}
  tabIndex={0}
/>
`;
